ZOOKEEPER-4333 QuorumSSLTest - testOCSP fails on JDK17

eolivelli opened the pull request:
https://issues.apache.org/jira/browse/ZOOKEEPER-4333

in JDK17 the OCSP request is sent in the URI and not inside the POST BODY

ztzg commented:
> kindly pinging @ztzg @maoling

Hi @eolivelli,

This is in my queue, and I am back in Munich.  I should be able to look into this on Saturday.

Cheers, -D

ztzg commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/QuorumSSLTest.java:
```diff
@@ -224,12 +228,24 @@ public OCSPHandler(X509Certificate revokedCert) {
             byte[] responseBytes;
             try {
+                String uri = httpExchange.getRequestURI().toString();
+                LOG.info("OCSP request: {} {} {}", httpExchange.getRequestMethod(), uri, httpExchange.getRequestHeaders().entrySet());
+                httpExchange.getRequestHeaders().entrySet().forEach((e) -> {
+                    LOG.info("OCSP request header: {} {}", e.getKey(), e.getValue());
+                });
```
This logs each header twice: once at the end of the `OCSP request: ` line, in "compressed form," and once per line in the rest.  We could perhaps drop the third format argument from the first line?

ztzg commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/QuorumSSLTest.java:
```diff
@@ -224,12 +228,24 @@ public OCSPHandler(X509Certificate revokedCert) {
                 byte[] requestBytes = new byte[10000];
-                request.read(requestBytes);
+                int len = request.read(requestBytes);
+                LOG.info("OCSP request size {}: {}", len, new String(requestBytes, StandardCharsets.UTF_8));
```
This dumps 10000 bytes of binary data in the log.  (Up to 10000 NUL bytes, if `len` is -1.)  May I suggest dropping the second format argument, as it is unreadable even when the OCSP request is passed in the body (the data does not seem to be UTF-8 encoded anyway)?

ztzg commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/QuorumSSLTest.java:
```diff
@@ -224,12 +228,24 @@ public OCSPHandler(X509Certificate revokedCert) {
+                LOG.info("OCSP request size {}: {}", len, new String(requestBytes, StandardCharsets.UTF_8));
 
+                if (len < 0) {
+                    String removedUriEncodiing = URLDecoder.decode(uri.substring(1), "utf-8");
```
Nit/typo: `removedUriEncoding`.

ztzg reviewed (changes requested):
Hi @eolivelli,

Sorry for the delay in reviewing this.

Generally LGTM, and tested under JDK 11 and 17-ea, with the expected results.

Most of my comments are trivial (particularly given that this is not core code), but I would like to remove the statement which accidentally dumps ~10kB of binary data to the log.  What do you think?

Besides that, the result looks ready to be merged.  Thank you for figuring it out!

Cheers, -D

eolivelli commented:
@ztzg thanks for your review.
I have addressed your comments
PTAL (and hopefully merge)
please cherry pick to branch-3.7 it is better that at least 3.7 passes the tests on modern JDKs

eolivelli commented:
@ztzg  CI passed

ztzg reviewed (approved):
LGTM!  Merging.

ztzg commented:
@eolivelli: This is now in `master` and `branch-3.7`.  Cheers!
